wordpress-mobile/WordPress-Android
# RI: Create expandable tags view

Parent Issue: #12028

**Review Instructions**

1. Review PR #12615
2. Make sure PR #12615 is merged to develop
3. Merge develop to this PR
4. Review this PR

**Description**

This PR creates custom expandable tags view for the discover post card.

Custom expandable tags view extends `ChipGroup` widget and adds two types of chips:
- Tag chip (corresponding to each tag)
- Overflow Indicator chip (added to the end of the `ChipGroup` as last child)

Overflow indicator chip is shown by hiding (`View.GONE`) all tag chips outside the visible area of the `ChipGroup`. 

**ChipGroup states**
Collapsed (single line) | Expanded (full height)
--- | ---
![collapsed](https://user-images.githubusercontent.com/1405144/89293912-547d2d80-d67c-11ea-9f89-593e1b5e3638.png) | <img width="320" alt="Screenshot 2020-08-04 at 5 57 50 PM" src="https://user-images.githubusercontent.com/1405144/89293811-27307f80-d67c-11ea-8ce2-fb76a5585dfd.png">

**Edge conditions**
- [Overflow indicator chip is hidden](https://github.com/wordpress-mobile/WordPress-Android/blob/c546b78ff7979456060ed6c2c7b3940c31d0e799/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderExpandableTagsView.kt#L128-L129) if there are no tag chips outside the visible area of the `ChipGroup`. 

- In the collapsed state, after hiding tag chips outside the visible area of the `ChipGroup`, there might be a case where the overflow indicator chip is still not shown. [Last visible tag chip is hidden](https://github.com/wordpress-mobile/WordPress-Android/blob/c546b78ff7979456060ed6c2c7b3940c31d0e799/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderExpandableTagsView.kt#L121-L125) in this case so that the overflow indicator chip gets shown.
<img width="320" alt="chips" src="https://user-images.githubusercontent.com/1405144/89292900-b2a91100-d67a-11ea-9f6b-1374dcae3712.png">

- Chip max width is set to a fraction (.75) of `ChipGroup` width. This is done to fit the overflow indicator chip in the collapsed state (single line) if the first chip title is very large.

- [Condition to find if tag chip is outside bounds](https://github.com/wordpress-mobile/WordPress-Android/blob/a5515b02e0225d5eb6cf2d7273ac3d59e2681765/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderExpandableTagsView.kt#L110-L118) is based on RTL and LTR layout direction of the `ChipGroup`.

**To test**

Note: Actual post tags retrieved in PR #12615 and set on expandable tag view in 63c9d8d

Prerequisite: RI FF is enabled

- Launch app (in portrait mode) and go to Reader tab
- Notice that 
    - tags are visible on the post card (only on discover tab)
    - correct tags (corresponding to post) are displayed
    - overflow chip is shown if tags cannot fit in the collapsed state
    - overflow chip displays correct no. of chips that are hidden
- Click overflow chip
- Notice that layout expands with a hide all option
- Click hide all option
- Notice that layout collapses, displaying an overflow indicator
- Switch to landscape mode
- Notice that expandable view adjusts to new width
- Click a tag 
- Notice that topic details page is opened for selected tag

**Note**
~~Used [requestLayout()](https://github.com/wordpress-mobile/WordPress-Android/blob/c546b78ff7979456060ed6c2c7b3940c31d0e799/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderExpandableTagsView.kt#L102) to get chip coordinates inside `isChipWithinBounds` else those are returned zero. I tried to override `FlowLayout.onMeasure` (`FlowLayout` is `ChipGroup` parent) but it can only be called from within the same library group (`com.google.android.material`). Let me know if you have suggestions to improve.~~

PR submission checklist:
